Abgabe Exercise #3

m-oehme opened the pull request.

Commit:
Moved client files into dedicated client folder

Commit:
added demo schema and some data for graphql

Commit:
Merge branch 'master' into develop

Commit:
fixed test run script

Commit:
added apollo client to access that sweet sweet data hmmm

Commit:
added dummy tests

Commit:
added some tests for apollo server

also split index.js into ore files

Commit:
added config for backend CI

Commit:
Merge branch 'master' into develop

updated develop branch after previous PR was approved

Commit:
fixed linter

Commit:
added update flag for jest

Commit:
did some styling and added a author

Commit:
added field to change current author

Commit:
adding an entry puts in correct author

Commit:
Update README.md

Commit:
fixed author assignment

Commit:
added mutation

vctosp committed:
Exercise 3, Task 2 finished

vctosp committed:
Exercise 3, Task 2 code

vctosp committed:
Defined query schema for tests

vctosp committed:
Exercise 3, Task 3, partly finished

vctosp committed:
Exercise 3, task 4 finished

vctosp committed:
adjusted ReadMe

Commit:
Update README.md

Commit:
Update README.md

vctosp committed:
changed readme titel

Commit:
Merge branch 'master' into develop

Commit:
added implementation for qraphql resolvers

needed some adjustment in testfile

Commit:
fixed tests

added author for todo but forgot to add it to mockup data

Commit:
added all calls to vue-apollo

Todo list is now in backend YAY

Commit:
edited test

there is a error in the tests

Commit:
fixed data in tests

because of vue-apollo we needed to set the data differently

Commit:
fixed jest config

jest types were not included

Commit:
made schema definition more modular

Commit:
typo

Commit:
moved backend tests to the resolvers

Commit:
defined login tests

Commit:
added login implementation

only simple login on backend

Commit:
added tests for signing up new user

Commit:
added signup implementaton

Commit:
throwing authentication error on failed authentication

Commit:
added componentn for login

Commit:
tokken handling in backend

Commit:
added load of authenticate todolist

Commit:
added signup

Commit:
checkmark

Commit:
fixed tests

Commit:
dfrfeefervrev

Commit:
changed edit mode bla

SyJuly commented on backend/src/schema/resolvers/login.js:
```diff
@@ -0,0 +1,46 @@
+      let user = userData.find(user => user.username === params.username);
+
+      if (user !== undefined) {
+        let token = jwt.sign(user, "supersecret");
```
Bei unserer App ist Sicherheit jetzt nicht so relevant, aber es wäre sicherlich gut, das supersecret aus einer separaten file zu lesen (welche dann in Realität nicht commited wird)

m-oehme replied:
das stimmt natürlich. War jetzt aber nicht prio. Haben wir der Einfachheit halber weg gelassen

SyJuly commented on backend/src/schema/resolvers/login.js:
```diff
@@ -0,0 +1,46 @@
+    login: (object, params) => {
+      let user = userData.find(user => user.username === params.username);
+
+      if (user !== undefined) {
```
sollte hier nicht auch überprüft werden, ob ein korrektes Password eingegeben wurde?

m-oehme replied:
damit wir uns nicht Verschlüsselung oder so beschäftigen müssen haben wir das Passwort weggelassen. Man "Loggt" sich nur mit einem Username ein.
Ist natürlich super unsicher 😉

SyJuly commented on backend/src/schema/resolvers/login.js:
```diff
@@ -0,0 +1,46 @@
+        };
+        userData.push(newUser);
+
+        let token = jwt.sign(newUser, "supersecret");
```
supersecret auslagern ;)

SyJuly commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+        username: userData[0].username
+      }
+    });
+    expect(res.data.login.isLoggedIn).toBeTruthy();
```
hier wird überprüft ob der boolean isLoggedIn gesetzt ist, nicht ob ein valides Token vom Server zurückkommt ("receiving token on login")

m-oehme replied:
ich war mir da nicht sicher wie man prüfen kann ob das Token valide ist. Das ist ja semi random 🤔

roschaefer replied:
Ah, ihr könnt einfach versuchen das token zu dekodieren, wenn ihr ganz sicher gehen wollt. Es wird einen Fehler werfen, wenn die Signatur nicht stimmt und das Token invalide ist. So machen wir's hier:

https://github.com/Human-Connection/Human-Connection/blob/master/backend/src/jwt/decode.spec.js#L32

SyJuly commented on backend/src/schema/types/Login.js:
```diff
@@ -0,0 +1,18 @@
+  type Login {
+    username: String
+    token: String
+    isLoggedIn: Boolean
```
wofür ist isLoggedIn wichtig? wenn ein valides token gesetzt ist, bedeutet das, dass der User eingeloggt ist

m-oehme replied:
Ich hatte mir das bei HumanConnection abgeschaut. Am ende ist mir auch aufgefallen das wir das gar nicht brauchen. Habs vergessen rauszunehmen.

Commit:
chaged authorization to use headers

Penomatikus reviewed (approved):
Nice that you have implemented an login to fulfill the task with the JWT implementation! Good job! 
I've viewed all your code and find it great!

Commit:
fixed token on frontend

roschaefer commented on README.md:
```diff
@@ -10,66 +10,24 @@ Development: [![Build Status](https://travis-ci.com/m-oehme/Systems-Development-
-Documentation: [https://github.com/vuejs/vue-test-utils](https://github.com/vuejs/vue-test-utils)
-
-**Required)** Installing:
+## Development Envorment Setup
```
```suggestion
## Development Environment Setup
```

roschaefer commented on backend/package.json:
```diff
@@ -0,0 +1,65 @@
+  "license": "ISC",
+  "dependencies": {
+    "apollo-server": "^2.9.9",
+    "apollo-server-testing": "~2.9.7",
```
`apollo-server-testing` shouldn't be part of `dependencies` but `devDependencies`.

roschaefer commented on backend/src/data.js:
```diff
@@ -0,0 +1,45 @@
+const books = [
```
At last, something other than todos :+1:

m-oehme replied:
this was actually not used. I removed it

roschaefer commented on backend/src/data.js:
```diff
@@ -0,0 +1,45 @@
+  }
+];
+
+module.exports.books = books;
```
If you setup babel, you can just say `export const books = [ /* ... */ ]`.

m-oehme replied:
we tried this but its not working for us. with ```export``` we have to use ```import``` and we always get an exception: ```SyntaxError: Cannot use import statement outside a module```

roschaefer replied:
That's exactly the kind of error message that you get when you didn't setup your transpiler correctly.

m-oehme replied:
Neither me nor Victor are JS devs. So we have no idea how to do that. We tried to google it with no success.

roschaefer replied:
https://github.com/m-oehme/Systems-Development-and-Frameworks/pull/6

Here you go :raised_hands: Please tell me if that works out for you.

m-oehme replied:
Works like charm 😁
thank you

roschaefer commented on backend/src/index.js:
```diff
@@ -0,0 +1,20 @@
+  typeDefs,
+  resolvers,
+  context: ({ req }) => {
+    const token = req.headers.authorization.replace("Bearer ", "");
```
You could decode the token here.

m-oehme replied:
we changed this in Exercise  5

roschaefer commented on backend/src/schema/resolvers/helloWorld.spec.js:
```diff
@@ -0,0 +1,30 @@
+    await expect(query({ query: GET_HELLO })).resolves.toMatchObject({
+      data: {
+        message: hello
+      }
```
```suggestion
      },
      errors: undefined
```

For extra assurance and nicer expectation failure messages.

roschaefer commented on backend/src/schema/resolvers/login.js:
```diff
@@ -0,0 +1,46 @@
+module.exports.LoginResolver = {
+  Mutation: {
+    login: (object, params) => {
+      let user = userData.find(user => user.username === params.username);
```
Knowing the username is enough to login? :thinking:

m-oehme replied:
Yes, we did this just to save some Dev time. In a real world scenario it would require a password at least that has to be encrypted.

roschaefer commented on backend/src/schema/resolvers/login.js:
```diff
@@ -0,0 +1,46 @@
+          isLoggedIn: true
+        };
+      } else {
+        throw new AuthenticationError(
```
I would probably throw `ValueError` or `ArgumentError` or something. It's not an authentication error.

m-oehme replied:
Its the login, so if authentication failed wouldn't that be an Auth error??

roschaefer replied:
I thought it's signup?

m-oehme replied:
My bad. You are correct, sir. 😑

m-oehme replied:
There is no such thing like ```ValueError`` or ```ArgumentError``` 
So I have to define it myself?

roschaefer replied:
Oh, well then just use `new Error` I would say.

m-oehme replied:
ok

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+const POST_LOGIN = gql`
+  mutation login($username: String) {
+    login(username: $username) {
+      isLoggedIn
```
Isn't it enough to have the `token` to know that you're logged in? :thinking:

m-oehme replied:
that is correct. We took HC implementation as reference and it has this. We probably change this in the future.

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+    let res = await mutate({
+      mutation: POST_LOGIN,
+      variables: {
+        username: userData[0].username
```
Now I have to lookup `"../../data"` in order to know what's going on. Why not hard-coding the `username` here? It's just one string isn't it?

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+        username: userData[0].username
+      }
+    });
+    expect(res.data.login.isLoggedIn).toBeTruthy();
```
```suggestion
    expect(res.data.login.isLoggedIn).toBe(expect.any(String));
```
would be a little more specific.

m-oehme replied:
this is also about ```isLoggedIn``` being useless. Changed it to:
```expect(res.data.login.token).toStrictEqual(expect.any(String))```

roschaefer replied:
Right, `isLoggedIn` wasn't the token but some boolean. I meant `token`.

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+});
+
+describe("Mutations", () => {
+  it("receiving token on login", async () => {
```
```suggestion
  describe('given valid login credentials', () => {
    it("responds with token", async () => {
```
Always use simple present and make the `it` part of the sentence. Describe what `it` does.

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+    expect(res.data.login.isLoggedIn).toBeTruthy();
+  });
+
+  it("error wrong username", async () => {
```
Same :point_up: 

```suggestion
  it("throws AuthenticationError", async () => {
```

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+        username: "Not A Human connected!"
+      }
+    });
+    expect(res).toMatchObject({
```
You can combine the two statements and get much better expectation failure messages.

```suggestion
    await expect(await mutate({ /* ... */ })).resolves.toMatchObject({ /* ... */ })
```

m-oehme replied:
I just tried this and the expectation failure message was much worse and not really readable.

with our implementation:
```
expect(received).toMatchObject(expected)

    - Expected
    + Received

      Object {
    -   "data": null,
    -   "errors": Array [
    -     Object {
    -       "message": "There is no such user, you fool!",
    +   "data": Object {
    +     "login": Object {
    +       "isLoggedIn": true,
    +       "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6Ik1heCIsImlhdCI6MTU3NzQ0MzUwNX0.x52nfDRLz_WWfn5SbAYnnJILQTU_wzimTqwU2vsUzeA",
    +     },
        },
    -   ],
    +   "errors": undefined,
      }
```

vs. Your Suggestion:
```
 expect(received).resolves.toMatchObject()

    Matcher error: received value must be a promise

    Received has type:  object
    Received has value: {"data": {"login": {"isLoggedIn": true, "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6Ik1heCIsImlhdCI6MTU3NzQ0MzYzMH0._Uxsav0hJDwon34RDa13k3tKX_dSddayAPsY-i-kY_U"}}, "errors": undefined, "extensions": undefined, "http": {"headers": {Symbol(map): {}}}}

      61 |           username: "Max"
      62 |         }
    > 63 |       })).resolves.toMatchObject({
         |                    ^
      64 |         data: null,
      65 |         errors: [{ message: "There is no such user, you fool!" }]
      66 |       });
```

roschaefer replied:
Oh yes, I made a copy+paste error. Of course, the `await expect(promise).resolves...` must be a promise, I forgot to remove the await by accident:
```js
    await expect(mutate({ /* ... */ })).resolves.toMatchObject({ /* ... */ })
```

m-oehme replied:
Ah ok. it works now.
But it return the same expectation  failure message as ours. So I don't really the the benefit except having a variable less.

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+    });
+  });
+
+  it("signup successful", async () => {
```
In this case you would expect that
```suggestion
  it("responds with a user", async () => {
```
and in another test case you could expect that

```suggestion
  it("creates a new user in the database", async () => {
```

roschaefer commented on backend/src/schema/resolvers/login.spec.js:
```diff
@@ -0,0 +1,80 @@
+    expect(res.data.signup.isLoggedIn).toBeTruthy();
+  });
+
+  it("signup failed", async () => {
```
```suggestion
  describe('user exists in database already', () => {
    beforeEach(() => { /* ... setup the user in the database */ })
    it("throws ArgumentError", async () => { /* ... */ })
  })
```

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+  },
+  Query: {
+    todos: (object, params, context) => {
+      const decoded = decodedToken(context.token);
```
If you would decode the token in the context hook you wouldn't have to do it in every resolver.

m-oehme replied:
We're doing this in upcoming Exercise 5

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+    }
+  },
+  Mutation: {
+    delToDo: (object, params, context) => {
```
```suggestion
    deleteTodo: (object, params, context) => {
```
I prefer not to use abbreviations.

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+    addToDo: (object, params, context) => {
+      const decoded = decodedToken(context.token);
+      var maxid = 0;
+      todoListData.map(obj => {
```
That's not a proper use of `map`! Use [reduce](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce).

Or take advantage of `Math.max()`
```js
Math.max(0, ...todoListData.map(t => t.id))
```

m-oehme replied:
I wouldn't call it "not proper", but reduce works better. I usually work with Kotlin which has much better Lambda implementations and some names are different. 😅

Could you explain the three dots in front of a map? ```...todoListData```

roschaefer replied:
`map` is always returning an array and the callback should return a value. So you're using a `map` instead of a `forEach`. `map` is supposed to *map* an array into another array.

The `...` notation is the destructuring assignment. Or call it "unpacking". https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
So my suggestion is to map the todoListData to an array of ids and call `Math.max` on the ids. The `0` is a placeholder for the edge case of an empty array.

`reduce` (my first suggestion) is returning a single item from a list.

m-oehme replied:
hm ok thank you

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+      });
+
+      todoListData.push({
+        id: maxid + 1,
```
I would simply use [uuid](https://www.npmjs.com/package/uuid) and never worry about unique ids anymore.

m-oehme replied:
not really necessary in our use-case. But I understand what you mean

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+    updateToDo: (object, params, context) => {
+      const decoded = decodedToken(context.token);
+      let index = todoListData.findIndex(todoData => todoData.id === params.id);
+      todoListData[index].text = params.text;
```
```suggestion
      todoListData[index] = { ...todoListData[index], ...params }
```

m-oehme replied:
see above 👆
what so the three dots mean??

roschaefer replied:
I'm unpacking two objects and merge the values into a new object here. The key-value pairs of the last object have precedence over the objects before, in this case `params` overrides `todoListData[index]`.

roschaefer commented on backend/src/schema/resolvers/todos.js:
```diff
@@ -0,0 +1,50 @@
+      let index = todoListData.findIndex(todoData => todoData.id === params.id);
+      todoListData[index].text = params.text;
+      todoListData[index].author.name = params.authorName;
+      return todoListData.filter(todo => todo.author.name === decoded.username);
```
I find it odd that `updateTodo` returns an array of todos instead of a single (the updated) object.

m-oehme replied:
this changes I the upcoming Exercise

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+
+const ADD_TODO = gql`
+  mutation addToDo($text: String!, $authorName: String!) {
+    addToDo(text: $text, authorName: $authorName) {
```
Very nice that you return an object with `author`. You could even go so far to give `addTodo` an input param of type `Author`. Your method call would look like this:

```gql
addToDo(text: $text, author: { name: $authorName }) {
```

m-oehme replied:
Oh I didn't know that's possible

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+  testServer.requestOptions = {
+    context() {
+      return {
+        token: token
```
Yep, and if you would decode the token in the `context` you would mock the authenticated user here.

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+  query = createTestClient(testServer).query;
+  mutate = createTestClient(testServer).mutate;
+
+  let res = await mutate({
```
```suggestion
  let { data: { login } } = await mutate({ /* ... */ )
  token = login
```

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+    });
+    expect(res).toMatchObject({
+      data: {
+        todos: todoListData.filter(todo => todo.author.name === "Max")
```
I would hard-code those values. It's too brittle otherwise.

m-oehme replied:
I agree. In next Exercise we're mocking the data so this won't be an issue. I'll just leave it as is in this PR

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+
+  it("delete todo", async () => {
+    const list = todoListData.filter(todo => todo.author.name === "Max");
+    list.splice(0, 1);
```
This is what I mean: I cannot imagine the value of `list` here anymore. If you would use a hard-coded value, that would make life so easy!

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+    list.splice(0, 1);
+    await expect(
+      mutate({ mutation: DEL_TODO, variables: { id: 1, token: token } })
+    ).resolves.toMatchObject({
```
Here you are using the promise syntax of `expect` :+1:

roschaefer commented on backend/src/schema/resolvers/todos.spec.js:
```diff
@@ -0,0 +1,171 @@
+
+  it("add todo", async () => {
+    const list = todoListData.filter(todo => todo.author.name === "Max");
+    list.push(newtodo);
```
Hard-code your `list` before and after you make your action if necessay. Honestly I don't know how it looks like. What are you testing here actually? `it("add todo"` :thinking: maybe "It adds a todo to the database"?

roschaefer commented on backend/src/schema/types/Todo.js:
```diff
@@ -0,0 +1,24 @@
+  type Mutation {
+    delToDo(id: ID): [Todo]
+    addToDo(text: String!, authorName: String!): [Todo]
+    updateToDo(id: ID!, text: String!, authorName: String!): [Todo]
```
Why on earth are those mutations returning `[Todo]` and not `Todo`? :thinking:

m-oehme replied:
this changes in the future

roschaefer commented on package.json:
```diff
@@ -3,47 +3,37 @@
-    "core-js": "^3.3.2",
-    "vue": "^2.6.10",
-    "vue-router": "^3.1.3"
+    "babel-core": "^7.0.0-bridge.0"
```
Isn't `babel-core` also a `devDependency`? :thinking: or do you need it for something at runtime?

m-oehme replied:
I have absolutely no idea

roschaefer commented on webapp/package.json:
```diff
@@ -0,0 +1,57 @@
+    "graphql-tag": "^2.10.1",
+    "vue": "^2.6.10",
+    "vue-apollo": "^3.0.0",
+    "vue-router": "^3.1.3"
```
Same here probably. I believe, if you have a plain SPA, there is no real dependency because you are not running any node process.

m-oehme replied:
Whats SPA?

roschaefer replied:
Single-page-application. VueJS is a single-page-application framework. I assume you build a couple of JS, HTML and CSS files and then just serve it with an ordinary file server. In that case, there is no node process running on the server. Probably you can put everything into the `devDependencies` section.

Once we introduce NuxtJS we will have runtime dependencies because we're running a node process on the server.

m-oehme replied:
I don't really know which dependency goes where. Maybe you can make an Overview in an upcoming class?

roschaefer commented on webapp/src/App.vue:
```diff
@@ -0,0 +1,49 @@
+  methods: {
+    setUser(user) {
+      this.user = user;
+      localStorage.setItem("token", user.token);
```
I recommend the [login helpers](https://github.com/nuxt-community/apollo-module#authentication) that come with `apollo` module for [NuxtJS](https://nuxtjs.org/).

m-oehme replied:
I am not going to implement NuxtJs

roschaefer commented on webapp/src/components/TodoItem.vue:
```diff
@@ -0,0 +1,81 @@
+      this.temp = this.todo;
+      this.isEditing = true;
+    },
+    delToDo: function() {
```
I would always try to avoid abbreviations.

roschaefer commented on webapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,215 @@
+      <!--      <div v-else class="head_right">-->
+      <!--        <button class="bigGreen" @click="saveAuthor">Save Author</button>-->
+      <!--        <input v-model="author.name" />-->
+      <!--      </div>-->
```
Remove commented code.

roschaefer commented on webapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,215 @@
+            }
+          })
+          .then(body => {
+            this.todoListData = body.data.updateToDo.slice();
```
With `vue-apollo` it shouldn't be necessary to update all that data manually. If you make your `id` of type `ID` vue-apollo is able to update the already existing item in the cache.

m-oehme replied:
I don't know how to do this when manually calling the Query

roschaefer commented on webapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,215 @@
+              authorName: this.user.username
+            }
+          })
+          .then(body => {
```
I would recommend `async/await` syntax.

m-oehme replied:
can the Methods from Vue component be ```async```??

roschaefer replied:
Sure, I think your build tool will take care of it.

m-oehme replied:
ok it looks a bit tidier

roschaefer commented on webapp/tests/unit/TodoList.spec.js:
```diff
@@ -0,0 +1,53 @@
+  it("$emits add todo", () => {
+    wrapper.find("#addTodo").trigger("click");
+
+    expect(wrapper.emitted()).toBeTruthy();
```
I would specify the name of the event.
```suggestion
    expect(wrapper.emitted('some-event')).toBe(true);
```

m-oehme replied:
there is no event here. it just calls a method

roschaefer commented on webapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,215 @@
+                }
+              }
+            `,
+            variables: {
```
It doesn't hurt to add `id` as unused variable for the if-branch of creating the todo.

So you can DRY your code like this:
```js
const mutation = !todo.id ? createMutation : updateMutation
const { id, text } = todo
const variables = { id, text, authorName: this.user.username }
const data = this.$apollo.mutate({mutation, variables})
```

m-oehme replied:
I don't like that. using ```todo.id``` makes it easier to identify where the ```id``` comes from.
Its not always about clean code. Readable code is important too 😉

roschaefer reviewed (approved):
![wow](https://media.giphy.com/media/3osxY7xVVSgMfAPZ16/giphy.gif)

Great! Well done! I haven't tried your code but it looks amazing! A couple of improvements I have for you though :point_down:

roschaefer commented:
And maybe squash/fixup your commits:

![Screenshot - 2019-12-27T012316 910](https://user-images.githubusercontent.com/2110676/71494414-6590d100-2847-11ea-82d0-6ecf1b7e4136.png)

roschaefer committed:
Setup nodemon and babel

This will do the following:

1. Watch your files, so if you make a change, your apollo-server will
automatically reload
2. Allow you to use `export default ..` statements

You can install all dependencies with this command:
```sh
npm install nodemon @babel/core @babel/node @babel/preset-env --save-dev
```

I followed this tutorial:
https://www.robinwieruch.de/minimal-node-js-babel-setup

And I copied some our
[configuration](https://github.com/Human-Connection/Human-Connection/blob/master/backend/babel.config.json).

Please tell me if this works out for you and I will send that solution
to the rest of the course.

Commit:
a lot of small changes suggested by @roschaefer

Commit:
Merge pull request #6 from m-oehme/setup_nodemon_and_babel

Setup nodemon and babel

Commit:
using import statement now

import wasnt working before
change in config made import and export possible
